greplogs/broken.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2022 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	_ "embed"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+//go:embed _embed/broken.go
+var brokenScript []byte
+
+// listBrokenBuilders returns the builders that are marked
+// as broken in golang.org/x/build/dashboard at HEAD.
+func listBrokenBuilders() (broken []string, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("identifying broken builders: %v", err)
+		}
+	}()
+
+	modDir, err := os.MkdirTemp("", "greplogs")
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		removeErr := os.RemoveAll(modDir)
+		if err == nil {
+			err = removeErr
+		}
+	}()
+
+	runCommand := func(name string, args ...string) ([]byte, error) {
+		cmd := exec.Command(name, args...)
+		cmd.Dir = modDir
+		cmd.Env = append(os.Environ(), "GO111MODULE=on", "PWD="+modDir)
+		cmd.Stderr = new(strings.Builder)
+
+		out, err := cmd.Output()
+		if err != nil {
+			return out, fmt.Errorf("%s: %w\nstderr:\n%s", strings.Join(cmd.Args, " "), err, cmd.Stderr)
+		}
+		return out, nil
+	}
+
+	_, err = runCommand("go", "mod", "init", "github.com/aclements/go-misc/greplogs/_embed")
```

Comment:
This was certainly... unexpected. I assume the idea here is that we actually need to get whatever's at HEAD to get the current broken builders and there's not, like, an RPC API we can hit to get this?

If I'm understanding right, this definitely could use a comment explaining the rationale here. (Also, could we instead add a dashboard API to get this?)

The other thing I wonder is how well-defined this is at all. This will get the list of broken builders *right now*, but greplogs is all about digging back in history. Doesn't this mean it could miss real failures that happened before a builder was marked broken, and surface expected failures after a builder stops being marked broken?

Reply:
> I assume the idea here is that we actually need to get whatever's at HEAD to get the current broken builders and there's not, like, an RPC API we can hit to get this?

Correct. And even if there was an RPC API, because the configuration is encoded as a Go package we would have to push a new version of some server every time we wanted to update it.

> If I'm understanding right, this definitely could use a comment explaining the rationale here.

Done!

> Doesn't this mean it could miss real failures that happened before a builder was marked broken, and surface expected failures after a builder stops being marked broken?

Yep! That's why we only use this if the `--triage` flag is set — we don't have a log of which builders were broken at which commits (and that would be fairly tedious to maintain), so instead we bias toward over-suppressing failures for builders that are currently broken.

(If triage falls too far behind and a builder is already fixed, a user can explicitly notch out additional previously-broken builders using the `--omit` flag.)

Reply:
> Correct. And even if there was an RPC API, because the configuration is encoded as a Go package we would have to push a new version of some server every time we wanted to update it.

Don't we have to do redeploy whenever the broken builder list changes anyway? As far as most people are concerned, the broken builders are whatever's grayed out on the dashboard. If the dashboard was also serving this list via an API, it would match up with whatever's grayed out.

(There's a bigger question here of whether it makes sense for this to be configured in a giant Go struct that can't track history and requires a redeploy on any changes, but that's for another time...)